### PR TITLE
refactor: [el-table] using sticky position refactor fixed column

### LINKED ISSUE #4345: [bug report] El table layout problem
### Element Plus version
1.2.0-beta.3

### OS/Browsers version
Chrome

### Vue version
3.2.21

### Reproduction Link
https://codepen.io/code-farmer-i/pen/GRvGLWW

### Steps to reproduce
给所有的列定义宽度后

### What is Expected?
布局正常

### What is actually happening?
布局错乱

### LINKED ISSUE #4550: [Bug Report] table use with fixed="right",  Summary not display
### Element Plus version
1.2.0-beta.3

### OS/Browsers version
mac/win10

### Vue version
3.2.23

### Reproduction Link
https://codepen.io/jkbopo/pen/PoKMBzV

image

### Steps to reproduce
The official website rendering out of the problem

### What is Expected?
table use with fixed="right"  show Summary 

### What is actually happening?
table use with fixed="right",  Summary not display

### LINKED ISSUE #4605: [Bug Report] Tabel using fixed and  expand, the hover is error
### Element Plus version
1.2.0-beta.4

### OS/Browsers version
chrome 96.0.4664.45

### Vue version
3.2.23

### Reproduction Link
https://codepen.io/liwenda/pen/XWeJpxd

### Steps to reproduce
open the first row, hover the son table row, the hover is error

### What is Expected?
open the first row, hover the son table row, the hover is true

### What is actually happening?
open the first row, hover the son table row, the hover is error

### PR DESCRIPTION
**Pull request by msidolphin**
re #4399

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: packages/components/table/src/table.vue [hunk 1]
```diff
@@ -139,115 +60,52 @@
...
-        />
-      </div>
+      <table-footer
+        :border="border"
+        :default-sort="defaultSort"
+        :store="store"
+        :style="{
+          width: layout.bodyWidth.value ? layout.bodyWidth.value + 'px' : '',
+        }"
+        :sum-text="sumText || t('el.table.sumText')"
```
Comment (JeremyWuuuuu) [inline packages/components/table/src/table.vue:105]: Same here:

```ts
const computedSumText = computed(() => props.sumText || t('el.table.sumText')
```

File: packages/components/table/src/util.ts [hunk 2]
```diff
@@ -370,3 +370,116 @@ export function createTablePopper(
...
+  } else {
+    if (after < store.states.fixedLeafColumnsLength.value) {
+      fixedLayout = 'left'
+    } else if (
+      start >=
+      columns.value.length - store.states.rightFixedLeafColumnsLength.value
+    ) {
+      fixedLayout = 'right'
+    }
+  }
```
Comment (JeremyWuuuuu) [inline packages/components/table/src/util.ts:413]: **Suggestion**  Can we do a switch case on this? I think we should avoid 3 branches if-else block.

File: packages/components/table/src/util.ts [hunk 3]
```diff
@@ -370,3 +370,116 @@ export function createTablePopper(
...
+      return s + (c.realWidth === null ? c.width : c.realWidth)
+    }, 0)
+  } else {
+    offset = columns
+      .slice(start + 1)
+      .reverse()
+      .reduce((acc, c) => {
+        return acc + (c.realWidth === null ? c.width : c.realWidth)
+      }, 0)
+  }
```
Comment (JeremyWuuuuu) [inline packages/components/table/src/util.ts:478]: For remove duplication, I think you should do:

```ts
const getOffset = (offset, column: Column) => {
  return offset + column.realWidth === null ? c.width : c.realWidth
}

Then in both reduce callback do:

```ts
if (isLeft) {
  offset = columns.slice(0, index).reduce(getOffset, 0)
} else {
  offset = columns.slice(start + 1).reverse().reduce(getOffset, 0)
}
```
```

File: packages/components/table/src/util.ts [hunk 4]
```diff
@@ -370,3 +370,116 @@ export function createTablePopper(
...
+  if (isLeft) {
+    styles.left = offset
+  } else {
+    styles.right = offset
+  }
```
Comment (JeremyWuuuuu) [inline packages/components/table/src/util.ts:483]: These can be assigned in the if-else block above without introducing a new if-else.

**Review by jw-foss — COMMENTED** [review]
@msidolphin I left some comments for you to consider, let me know if you have any questions 😃

File: packages/components/table/src/table-footer/index.ts [hunk 5]
```diff
@@ -42,24 +42,26 @@ export default defineComponent({
     },
   },
   setup(props) {
-    const { hasGutter, getRowClasses, columns } = useStyle(
-      props as TableFooter<DefaultRow>
-    )
+    const { hasGutter, getRowClasses, getRowStyles, columns, gutterWidth } =
```
[the remark this answers is not in the document]
Comment (msidolphin) [reply packages/components/table/src/table-footer/index.ts]: There are some differences between header, body and footer, so I think it is more appropriate in the `useStyle`

File: packages/components/table/src/table-footer/index.ts [hunk 6]
```diff
@@ -117,19 +124,8 @@ export default defineComponent({
...
-                        : '',
-                    ],
-                    style: this.getRowStyles(
-                      column,
-                      cellIndex,
-                      hasGutter,
-                      gutterWidth
-                    ),
+                    class: getCellClasses(columns, cellIndex, hasGutter),
+                    style: getCellStyles(column, cellIndex, hasGutter),
```
Comment (JeremyWuuuuu) [inline packages/components/table/src/table-footer/index.ts:128]: **Ideas:** I just came up with an idea that `getCellClasses` and `getCellStyles` are computed every single time, and the value might not be the same, not familiar with how `Vue` compares them(maybe we should dig it in), but we can cache this value

We can do this later.

**Review by jw-foss — APPROVED** [review]
This change LGTM, approved.

**Pull request merged**